Classloader changed in quarkus 2.1.0 (regression from 2.0.1) for optaplanner/drools with smallrye (but not resteasy)

### Describe the bug

The drools ANC compiler fails in 2.1.0 (not in 2.0.1) on optaplanner's activemq-quarkus quickstart
because the classloader changed.

Before (2.0.1):
![image](https://user-images.githubusercontent.com/176880/127677239-251cfa44-3786-4afc-873c-d07aeec03ab8.png)

After (2.1.0):
![image](https://user-images.githubusercontent.com/176880/127677319-e69f3f6e-627d-4518-b7f4-d4b18390f4e2.png)

The strange thing is, that it doesn't fail for any of the other quickstarts (those use resteasy and sometimes hibernate, this one uses smallrye).
Comparing the dependency tree between 2.0.1 and 2.1.0 shows nothing besides the quarkus version change that could explain this.

### Expected behavior

A successful test with the reproducer below.

### Actual behavior

```
2021-07-30 16:29:06,337 WARN  [org.acm.sch.mes.TimeTableMessagingHandler] (ForkJoinPool.commonPool-worker-3) Error during processing a solver request (1).: org.drools.ancompiler.CouldNotCreateAlphaNetworkCompilerException: Cannot create Compiled Alpha Network
        at org.drools.ancompiler.CompiledNetworkSource.newCompiledNetworkInstance(CompiledNetworkSource.java:80)
        at org.drools.ancompiler.CompiledNetworkSource.createInstanceAndSet(CompiledNetworkSource.java:66)
        at org.drools.ancompiler.KieBaseUpdaterANC.lambda$inMemoryUpdate$0(KieBaseUpdaterANC.java:72)
        at java.base/java.util.HashMap$Values.forEach(HashMap.java:976)
        at org.drools.ancompiler.KieBaseUpdaterANC.inMemoryUpdate(KieBaseUpdaterANC.java:70)
        at org.drools.ancompiler.KieBaseUpdaterANC.run(KieBaseUpdaterANC.java:52)
        at org.drools.ancompiler.KieBaseUpdaterANC.generateAndSetInMemoryANC(KieBaseUpdaterANC.java:99)
        at org.optaplanner.core.impl.score.stream.drools.DroolsConstraintSessionFactory.buildKieBaseFromModel(DroolsConstraintSessionFactory.java:85)
        at org.optaplanner.core.impl.score.stream.drools.DroolsConstraintSessionFactory.buildSession(DroolsConstraintSessionFactory.java:126)
        at org.optaplanner.core.impl.score.director.stream.DroolsConstraintStreamScoreDirectorFactory.newConstraintStreamingSession(DroolsConstraintStreamScoreDirectorFactory.java:50)
        at org.optaplanner.core.impl.score.director.stream.DroolsConstraintStreamScoreDirector.resetConstraintStreamingSession(DroolsConstraintStreamScoreDirector.java:75)
        at org.optaplanner.core.impl.score.director.stream.DroolsConstraintStreamScoreDirector.setWorkingSolution(DroolsConstraintStreamScoreDirector.java:67)
        at org.optaplanner.core.impl.solver.scope.SolverScope.setWorkingSolutionFromBestSolution(SolverScope.java:215)
        at org.optaplanner.core.impl.solver.AbstractSolver.solvingStarted(AbstractSolver.java:80)
        at org.optaplanner.core.impl.solver.DefaultSolver.solvingStarted(DefaultSolver.java:200)
        at org.optaplanner.core.impl.solver.DefaultSolver.solve(DefaultSolver.java:174)
        at org.acme.schooltimetabling.messaging.TimeTableMessagingHandler.lambda$solve$0(TimeTableMessagingHandler.java:79)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1736)
        at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1728)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
        at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)
Caused by: java.lang.NoSuchMethodException: org.drools.ancompiler.Compiledorg_acme_schooltimetabling_domain_LessonNetwork31008790748.<init>(org.drools.core.spi.InternalReadAccessor, java.util.Map)
        at java.base/java.lang.Class.getConstructor0(Class.java:3349)
        at java.base/java.lang.Class.getDeclaredConstructor(Class.java:2553)
        at org.drools.ancompiler.CompiledNetworkSource.newCompiledNetworkInstance(CompiledNetworkSource.java:77)
        ... 23 more

```

### How to Reproduce?

```
git clone https://github.com/kiegroup/optaplanner-quickstarts.git
cd optaplanner-quickstarts
git checkout development
cd technology/java-activemq-quarkus
gedit pom.xml // Switch it to 2.1.0.Final
mvn clean install -DskipTests
cd solver
mvn test -Dtest=TimeTableMessagingHandlerTest#solve
```

### Output of `uname -a` or `ver`

_No response_

### Output of `java -version`

_No response_

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

_No response_

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

_No response_

### Additional information

Reproduced on multiple machines (Radovan, Geoffrey, Luca).

--- a/core/deployment/src/main/java/io/quarkus/runner/bootstrap/ForkJoinClassLoading.java
+++ b/core/deployment/src/main/java/io/quarkus/runner/bootstrap/ForkJoinClassLoading.java
@@ -20,42 +20,31 @@ public class ForkJoinClassLoading {
      * Really we should just not use the common pool at all.
      */
     public static void setForkJoinClassLoader(ClassLoader classLoader) {
-        final ForkJoinPool commonPool = ForkJoinPool.commonPool();
-
-        final int activeThreadCount = commonPool.getActiveThreadCount();
-
-        //The following approach is rather heavy, but there's good chances that there are actually no active threads,
-        //so skip it in this case.
-        //Additionally, if we can skip this this will prevent us from actually forcing to start all threads in the pool.
-        if (activeThreadCount > 0) {
-            //TODO we could theoretically improve this by using the system property java.util.concurrent.ForkJoinPool.common.threadFactory
-            //and force the common pool to register created threads so that we can reach and reset them.
-            CountDownLatch allDone = new CountDownLatch(ForkJoinPool.getCommonPoolParallelism());
-            CountDownLatch taskRelease = new CountDownLatch(1);
-            for (int i = 0; i < ForkJoinPool.getCommonPoolParallelism(); ++i) {
-                commonPool.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        Thread.currentThread().setContextClassLoader(classLoader);
-                        allDone.countDown();
-                        try {
-                            taskRelease.await();
-                        } catch (InterruptedException e) {
-                            log.error("Failed to set fork join ClassLoader", e);
-                        }
+        CountDownLatch allDone = new CountDownLatch(ForkJoinPool.getCommonPoolParallelism());
+        CountDownLatch taskRelease = new CountDownLatch(1);
+        for (int i = 0; i < ForkJoinPool.getCommonPoolParallelism(); ++i) {
+            ForkJoinPool.commonPool().execute(new Runnable() {
+                @Override
+                public void run() {
+                    Thread.currentThread().setContextClassLoader(classLoader);
+                    allDone.countDown();
+                    try {
+                        taskRelease.await();
+                    } catch (InterruptedException e) {
+                        log.error("Failed to set fork join ClassLoader", e);
                     }
-                });
-            }
-            try {
-                if (!allDone.await(1, TimeUnit.SECONDS)) {
-                    log.error(
-                            "Timed out trying to set fork join ClassLoader, this should never happen unless something has tied up a fork join thread before the app launched");
                 }
-            } catch (InterruptedException e) {
-                log.error("Failed to set fork join ClassLoader", e);
-            } finally {
-                taskRelease.countDown();
+            });
+        }
+        try {
+            if (!allDone.await(1, TimeUnit.SECONDS)) {
+                log.error(
+                        "Timed out trying to set fork join ClassLoader, this should never happen unless something has tied up a fork join thread before the app launched");
             }
+        } catch (InterruptedException e) {
+            log.error("Failed to set fork join ClassLoader", e);
+        } finally {
+            taskRelease.countDown();
         }
     }
 }

